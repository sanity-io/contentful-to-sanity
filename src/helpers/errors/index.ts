export * from './ContentfulNoDefaultLocaleError'
export * from './ContentfulNoLocalesError'
export * from './ContentfulParamsMissingError'
export * from './OutputDirectoryNotEmptyError'
export * from './SanityApiError'
export * from './SanityDatasetNotFoundError'
export * from './SanityNotAuthenticatedError'
export * from './SanityProjectNotFoundError'
