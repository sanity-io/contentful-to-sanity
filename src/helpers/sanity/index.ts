export * from './defaultSanityClient'
export * from './deafultSanityHttpClient'
export * from './createProject'
export * from './writeSingleSanitySchema'
export * from './writeRootSanitySchema'
export * from './bootstrapStudio'
export * from './serializeRuleSpecToCode'
