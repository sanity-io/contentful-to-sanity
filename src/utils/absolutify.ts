import path from 'node:path'

import {expandHome} from './expandHome'

export function absolutify(dir: string): string {
  const pathName = expandHome(dir)
  return path.isAbsolute(pathName) ? pathName : path.resolve(process.cwd(), pathName)
}
