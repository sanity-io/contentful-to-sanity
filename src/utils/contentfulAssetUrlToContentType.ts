import type {AssetProps} from 'contentful-management'

import {prefixUrl} from './contentfulLinkToSanityReference'
export type ContentfulAssetFields = Pick<AssetProps, 'fields'>[]

export function contentfulAssetUrlToContentType(
  url: string,
  assets: ContentfulAssetFields,
  localeId: string,
) {
  const asset = assets.find((a) => {
    if (!a.fields.file) {
      return false
    }
    const assetUrl = a.fields.file[localeId].url
    return assetUrl && prefixUrl(assetUrl) === prefixUrl(url)
  })
  if (asset) {
    return asset?.fields.file[localeId].contentType
  }

  // If we can't find the asset, we'll try to guess the content type from the file extension
  const extension = url.split('.').pop()
  if (extension) {
    return `image/${extension}`
  }

  return undefined
}
