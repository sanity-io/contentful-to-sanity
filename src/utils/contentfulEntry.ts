import {type EntryProps} from 'contentful-management'

export function isDraft(entry: EntryProps<Record<string, Record<string, any>>>): boolean {
  return entry.sys.publishedVersion === undefined
}

export function isChanged(entry: EntryProps<Record<string, Record<string, any>>>): boolean {
  return !!entry.sys.publishedVersion && entry.sys.version >= entry.sys.publishedVersion + 2
}

export function isPublished(entry: EntryProps<Record<string, Record<string, any>>>): boolean {
  return !!entry.sys.publishedVersion && entry.sys.version == entry.sys.publishedVersion + 1
}
