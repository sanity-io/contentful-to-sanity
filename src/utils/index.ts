export * from './absolutify'
export * from './contentfulEntryToSanityObject'
export * from './contentfulFieldItemToSanityOfType'
export * from './contentfulFieldToSanityField'
export * from './contentfulLinkToSanityReference'
export * from './contentfulTypeToSanitySchema'
export * from './createIntlFields'
export * from './error'
export * from './expandHome'
export * from './extractContentfulRichTextFieldParameters'
export * from './extractValidationRulesFromContentfulField'
export * from './findEditorControlForField'
export * from './generateKey'
export * from './markdownToBlocks'
export * from './objectIsContentfulLink'
export * from './objectIsContentfulLocation'
export * from './objectIsContentfulRichText'
