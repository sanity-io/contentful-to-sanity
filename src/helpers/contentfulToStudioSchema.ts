import Case from 'case'
import {stringify} from 'javascript-stringify'
import compact from 'just-compact'
import prettier from 'prettier'

import {stringFieldSchemaFactory} from '../helpers/sanity/fieldSchemaFactories'
import {ContentfulExport, SanityDocumentSchema, SanityObjectSchema} from '../types'
import {contentfulTypeToSanitySchema} from '../utils/contentfulTypeToSanitySchema'
import {ContentfulNoDefaultLocaleError} from './errors/ContentfulNoDefaultLocaleError'
import {serializeRuleSpecToCode} from './sanity/serializeRuleSpecToCode'

export async function contentfulToStudioSchema(
  data: ContentfulExport,
  opts: {
    typescript: boolean
    intlMode: 'single' | 'multiple'
    keepMarkdown: boolean
    filepath: string
  },
): Promise<string> {
  const schemas: (SanityDocumentSchema | SanityObjectSchema)[] = compact(
    data.contentTypes!.map((type) => data && contentfulTypeToSanitySchema(type, data, opts)),
  )

  // add object break schema
  if (
    data.editorInterfaces?.some(
      (editor) => editor.controls?.some((ctrl) => ctrl.widgetId === 'richTextEditor'),
    )
  ) {
    const alreadyHasBreakSchema = schemas.some(({name}) => name === 'break')
    if (alreadyHasBreakSchema) {
      // eslint-disable-next-line no-console
      console.warn(
        'Found user-defined content model called "break". Be aware this could result in broken portable text',
      )
    }

    if (!alreadyHasBreakSchema) {
      schemas.push({
        name: 'break',
        title: 'Break',
        type: 'object',
        fields: [
          // @ts-expect-error - @TODO fix up the schema definitions
          stringFieldSchemaFactory('style')
            .options({
              list: [
                {title: 'Line break', value: 'lineBreak'},
                {title: 'Read more', value: 'readMore'},
              ],
            })
            .build(),
        ],
      })
    }
  }

  // add tag schema
  if (data.tags?.length) {
    const alreadyHasTagSchema = schemas.some(({name}) => name === 'tag')
    if (alreadyHasTagSchema) {
      // eslint-disable-next-line no-console
      console.warn(
        'Found user-defined content model called "tag". Please review manually as this could conflict with the tags data import from contentful',
      )
    }

    if (!alreadyHasTagSchema) {
      schemas.push({
        name: 'tag',
        title: 'Tag',
        type: 'document',
        fields: [
          // @ts-expect-error - @TODO fix up the schema definitions
          stringFieldSchemaFactory('name')
            .title('Name')
            .validation([{flag: 'presence', constraint: 'required'}])
            .build(),
        ],
      })
    }
  }

  const useMultiLocale = opts.intlMode === 'multiple'
  const defaultLocale = data.locales?.find((locale) => Boolean(locale.default))
  if (!defaultLocale) {
    throw new ContentfulNoDefaultLocaleError()
  }

  const allSchemaTypes = schemas.map((schema) => {
    const identifier = `${Case.camel(schema.name)}Type`
    const stringifiedDefinition = stringify(schema, (value, space, next, key) => {
      if (key === 'validation') {
        if (Array.isArray(value) && value.length > 0) {
          return `Rule => Rule.${value.map((r) => serializeRuleSpecToCode(r)).join('.')}`
        }

        return
      }

      if (opts.typescript && key === 'fields') {
        if (Array.isArray(value) && value.length > 0) {
          return `[${value.map((v, k) => `defineField(${next(v, k)})`).join(',')}]`
        }

        return
      }

      if (key === 'initialValue') {
        if (useMultiLocale) {
          // eslint-disable-next-line no-warning-comments
          // @TODO
        } else {
          return next(value[defaultLocale.code], key)
        }
      }

      return next(value, key)
    })
    const definition = `
    export const ${schema.name}Type = ${
      opts.typescript ? `defineType(${stringifiedDefinition})` : stringifiedDefinition
    };`
    return {identifier, definition}
  })

  const typesConcatList = allSchemaTypes.map((t) => `  ${t.identifier},`).join('\n')

  const result = `// generated by contentful-to-sanity
${
  opts.typescript
    ? 'import {defineField, defineType, type SchemaTypeDefinition} from "sanity";'
    : ''
}

${
  opts.typescript
    ? "const validateIn = (values: (string | number)[], value: any) => values.includes(value) ? true : `Value must be one of ${values.join(', ')}`"
    : "const validateIn = (values, value) => values.includes(value) ? true : `Value must be one of ${values.join(', ')}`"
}

${allSchemaTypes.map((t) => t.definition).join('\n\n')}


export const types = [
  ${typesConcatList}
] ${opts.typescript ? 'satisfies SchemaTypeDefinition[];' : ''}
  `
  return await formatWithPrettier(opts.filepath, result)
}

const formatWithPrettier = async (filepath: string, content: string) => {
  const config = await prettier.resolveConfig(filepath)

  return prettier.format(content, {
    ...config,
    filepath,
  })
}
