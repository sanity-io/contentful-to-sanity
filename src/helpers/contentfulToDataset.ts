import {SanityDocument} from '@sanity/client'
import compact from 'just-compact'
import invariant from 'tiny-invariant'

import {type ContentfulExport} from '../types'
import {isPublished} from '../utils/contentfulEntry'
import {contentfulEntryToSanityObject} from '../utils/contentfulEntryToSanityObject'
import {ContentfulNoDefaultLocaleError} from './errors/ContentfulNoDefaultLocaleError'
import {ContentfulNoLocalesError} from './errors/ContentfulNoLocalesError'

export async function contentfulToDataset(
  exports: {
    drafts: ContentfulExport
    published?: ContentfulExport
  },
  opts: {
    intlMode: 'single' | 'multiple'
    weakRefs: boolean
    intlIdStructure: 'subpath' | 'delimiter'
    keepMarkdown: boolean
    locale: string | undefined
  },
): Promise<string> {
  const parseExport = (data: ContentfulExport, isContentDeliveryAPIExport = false) => {
    const useMultiLocale = opts.intlMode === 'multiple'
    const defaultLocale = data.locales?.find((locale) => Boolean(locale.default))
    if (!defaultLocale) {
      throw new ContentfulNoDefaultLocaleError()
    }
    const localesToImport = (
      useMultiLocale
        ? (data.locales ?? []).map(({code}) => code)
        : opts.locale
        ? [opts.locale]
        : compact([defaultLocale?.code])
    ).filter((code) => data?.locales?.some((locale) => locale.code === code))

    if (!defaultLocale) {
      throw new ContentfulNoDefaultLocaleError()
    }

    if (localesToImport.length === 0) {
      throw new ContentfulNoLocalesError()
    }

    invariant(data.entries, 'Expected data.entries to be defined')
    invariant(data.entries.length > 0, 'Expected data.entries to be defined')

    const importableEntries: Set<SanityDocument> = new Set()
    for (const entry of data.entries) {
      for (const locale of localesToImport) {
        // We only want to import published entries if we're importing from the Content Delivery API
        if (!isContentDeliveryAPIExport && isPublished(entry)) continue
        const id = isContentDeliveryAPIExport
          ? entry.sys.id
          : isPublished(entry)
          ? entry.sys.id
          : `drafts.${entry.sys.id}`
        importableEntries.add(
          contentfulEntryToSanityObject(id, entry, locale, data, {
            useMultiLocale,
            idStructure: opts.intlIdStructure,
            defaultLocale: defaultLocale.code,
            supportedLocales: localesToImport,
            keepMarkdown: opts.keepMarkdown,
            weakRefs: opts.weakRefs,
          }),
        )
      }
    }

    return [...importableEntries].map((entry) => JSON.stringify(entry)).join('\n')
  }

  return [
    exports.drafts ? parseExport(exports.drafts, false) : '',
    exports.published ? parseExport(exports.published, true) : '',
  ]
    .filter((line) => line.length)
    .join('\n')
}
