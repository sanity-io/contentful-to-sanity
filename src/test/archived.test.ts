import type {ContentTypeProps} from 'contentful-management'
import {describe, expect, test} from 'vitest'

import {contentfulToDataset} from '../helpers/contentfulToDataset'
import {contentfulToStudioSchema} from '../helpers/contentfulToStudioSchema'
import {contentfulTypeToSanitySchema} from '../utils'
import {parse} from './helpers'

const options = {
  filepath: './fixtures/schema.ts',
  intlMode: 'single',
  keepMarkdown: false,
} as const

describe('Archived content entities', async () => {
  test('generated sanity document types have a hidden contentFulArchived field', async () => {
    const {default: data} = await import('./fixtures/archived.json')

    expect(data.contentTypes.length).toBe(2)

    const sanityContentTypes = []
    for (const contentType of data.contentTypes) {
      sanityContentTypes.push(
        contentfulTypeToSanitySchema(contentType as ContentTypeProps, data as any, {
          keepMarkdown: true,
        }),
      )
    }

    expect(sanityContentTypes.length).toBe(2)

    sanityContentTypes.forEach((doc) => {
      expect(doc.fields).toEqual(
        expect.arrayContaining([
          expect.objectContaining({
            name: 'contentfulArchived',
            type: 'boolean',
            readOnly: true,
          }),
        ]),
      )
    })
  })

  test('generated document schemas have readOnly function that reads contentfulArchived field, TS', async () => {
    const {default: data} = await import('./fixtures/archived.json')
    expect(
      await contentfulToStudioSchema(data as any, {...options, typescript: true}),
    ).toMatchSnapshot()
  })

  test('generated document schemas have readOnly function that reads contentfulArchived field, JS', async () => {
    const {default: data} = await import('./fixtures/archived.json')
    expect(
      await contentfulToStudioSchema(data as any, {...options, typescript: false}),
    ).toMatchSnapshot()
  })

  test('dataset action sets contentfulArchived field to "true" if the content entity is archived', async () => {
    const {default: drafts} = await import('./fixtures/archived.json')
    const {default: published} = await import('./fixtures/archived.published.json')
    const docs = (
      await contentfulToDataset(
        {drafts: drafts as any, published: published as any},
        {
          locale: 'en-US',
          weakRefs: false,
          intlMode: 'single',
          intlIdStructure: 'delimiter',
          keepMarkdown: true,
        },
      )
    )
      .split('\n')
      .map(parse)

    const archivedPost = docs.find(doc => doc._id === '2iacrOBkuYMOQyDWPlEJLc')
    expect(archivedPost).toBeDefined()
    expect(archivedPost).toHaveProperty('contentfulArchived', true)
  })
})
