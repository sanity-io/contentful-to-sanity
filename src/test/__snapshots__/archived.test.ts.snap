// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Archived content entities > generated document schemas have readOnly function that reads contentfulArchived field, JS 1`] = `
"// generated by contentful-to-sanity

export const personType = {
  type: 'document',
  name: 'person',
  title: 'Person',
  description: 'A human being',
  fields: [
    {
      name: 'name',
      type: 'string',
      title: 'name',
      hidden: false,
      description: 'Persons given, or preferred name',
      validation: (Rule) => Rule.required(),
    },
    {
      name: 'profilePicture',
      type: 'file',
      title: 'profilePicture',
      hidden: false,
      validation: (Rule) => Rule.required(),
    },
    {
      name: 'biography',
      type: 'array',
      of: [
        {
          type: 'block',
          styles: [{title: 'Quote', value: 'blockquote'}],
          lists: [
            {title: 'Bullet', value: 'bullet'},
            {title: 'Numbered', value: 'number'},
          ],
          marks: {
            decorators: [
              {title: 'Strong', value: 'strong'},
              {title: 'Emphasis', value: 'em'},
              {title: 'Underline', value: 'underline'},
            ],
            annotations: [
              {
                type: 'object',
                name: 'link',
                title: 'url',
                fields: [
                  {
                    type: 'string',
                    name: 'href',
                    title: 'URL',
                    validation: (Rule) => Rule.required(),
                  },
                  {
                    type: 'string',
                    name: 'target',
                    title: 'Target',
                    options: {
                      list: [
                        {value: '_blank', title: 'Blank'},
                        {value: '_parent', title: 'Parent'},
                      ],
                    },
                  },
                ],
              },
              {
                type: 'object',
                name: 'internalLink',
                title: 'Internal link',
                fields: [
                  {type: 'reference', name: 'reference', to: [{type: 'person'}, {type: 'post'}]},
                ],
              },
            ],
          },
          of: [{type: 'reference', title: 'Reference', to: [{type: 'person'}, {type: 'post'}]}],
        },
        {type: 'reference', title: 'Reference', to: [{type: 'person'}, {type: 'post'}]},
        {type: 'image'},
        {type: 'file'},
      ],
      title: 'biography',
      hidden: false,
    },
    {
      name: 'friend',
      type: 'reference',
      title: 'friend',
      hidden: false,
      to: [{type: 'person'}, {type: 'post'}],
    },
    {
      name: 'friends',
      type: 'array',
      of: [{type: 'reference', to: [{type: 'person'}, {type: 'post'}]}],
      title: 'friends',
      hidden: false,
    },
    {
      type: 'boolean',
      description:
        'If this document was archived on Contentful at the time of export, the document will be in a read-only state.',
      name: 'contentfulArchived',
      readOnly: true,
    },
  ],
  preview: {select: {title: 'name'}},
  readOnly: ({document}) => document?.contentfulArchived === true,
}

export const postType = {
  type: 'document',
  name: 'post',
  title: 'Post',
  description: '',
  fields: [
    {name: 'title', type: 'string', title: 'title', hidden: false},
    {
      name: 'author',
      type: 'reference',
      title: 'author',
      hidden: false,
      to: [{type: 'person'}, {type: 'post'}],
    },
    {
      name: 'body',
      type: 'array',
      of: [
        {
          type: 'block',
          styles: [
            {title: 'Heading 1', value: 'h1'},
            {title: 'Heading 2', value: 'h2'},
            {title: 'Heading 3', value: 'h3'},
            {title: 'Heading 4', value: 'h4'},
            {title: 'Heading 5', value: 'h5'},
            {title: 'Heading 6', value: 'h6'},
            {title: 'Quote', value: 'blockquote'},
          ],
          lists: [
            {title: 'Bullet', value: 'bullet'},
            {title: 'Numbered', value: 'number'},
          ],
          marks: {
            decorators: [
              {title: 'Strong', value: 'strong'},
              {title: 'Emphasis', value: 'em'},
              {title: 'Underline', value: 'underline'},
              {title: 'Code', value: 'pre'},
            ],
            annotations: [
              {
                type: 'object',
                name: 'link',
                title: 'url',
                fields: [
                  {
                    type: 'string',
                    name: 'href',
                    title: 'URL',
                    validation: (Rule) => Rule.required(),
                  },
                  {
                    type: 'string',
                    name: 'target',
                    title: 'Target',
                    options: {
                      list: [
                        {value: '_blank', title: 'Blank'},
                        {value: '_parent', title: 'Parent'},
                      ],
                    },
                  },
                ],
              },
              {
                type: 'object',
                name: 'internalLink',
                title: 'Internal link',
                fields: [
                  {type: 'reference', name: 'reference', to: [{type: 'person'}, {type: 'post'}]},
                ],
              },
              {type: 'image'},
              {type: 'file'},
            ],
          },
          of: [{type: 'reference', title: 'Reference', to: [{type: 'person'}, {type: 'post'}]}],
        },
        {type: 'reference', title: 'Reference', to: [{type: 'person'}, {type: 'post'}]},
        {type: 'image'},
        {type: 'file'},
        {type: 'break'},
      ],
      title: 'body',
      hidden: false,
    },
    {
      type: 'boolean',
      description:
        'If this document was archived on Contentful at the time of export, the document will be in a read-only state.',
      name: 'contentfulArchived',
      readOnly: true,
    },
  ],
  preview: {select: {title: 'title'}},
  readOnly: ({document}) => document?.contentfulArchived === true,
}

export const breakType = {
  name: 'break',
  title: 'Break',
  type: 'object',
  fields: [
    {
      name: 'style',
      type: 'string',
      options: {
        list: [
          {title: 'Line break', value: 'lineBreak'},
          {title: 'Read more', value: 'readMore'},
        ],
      },
    },
  ],
}

export const types = [personType, postType, breakType]
"
`;

exports[`Archived content entities > generated document schemas have readOnly function that reads contentfulArchived field, TS 1`] = `
"// generated by contentful-to-sanity
import {defineField, defineType, type SchemaTypeDefinition} from 'sanity'

export const personType = defineType({
  type: 'document',
  name: 'person',
  title: 'Person',
  description: 'A human being',
  fields: [
    defineField({
      name: 'name',
      type: 'string',
      title: 'name',
      hidden: false,
      description: 'Persons given, or preferred name',
      validation: (Rule) => Rule.required(),
    }),
    defineField({
      name: 'profilePicture',
      type: 'file',
      title: 'profilePicture',
      hidden: false,
      validation: (Rule) => Rule.required(),
    }),
    defineField({
      name: 'biography',
      type: 'array',
      of: [
        {
          type: 'block',
          styles: [{title: 'Quote', value: 'blockquote'}],
          lists: [
            {title: 'Bullet', value: 'bullet'},
            {title: 'Numbered', value: 'number'},
          ],
          marks: {
            decorators: [
              {title: 'Strong', value: 'strong'},
              {title: 'Emphasis', value: 'em'},
              {title: 'Underline', value: 'underline'},
            ],
            annotations: [
              {
                type: 'object',
                name: 'link',
                title: 'url',
                fields: [
                  defineField({
                    type: 'string',
                    name: 'href',
                    title: 'URL',
                    validation: (Rule) => Rule.required(),
                  }),
                  defineField({
                    type: 'string',
                    name: 'target',
                    title: 'Target',
                    options: {
                      list: [
                        {value: '_blank', title: 'Blank'},
                        {value: '_parent', title: 'Parent'},
                      ],
                    },
                  }),
                ],
              },
              {
                type: 'object',
                name: 'internalLink',
                title: 'Internal link',
                fields: [
                  defineField({
                    type: 'reference',
                    name: 'reference',
                    to: [{type: 'person'}, {type: 'post'}],
                  }),
                ],
              },
            ],
          },
          of: [{type: 'reference', title: 'Reference', to: [{type: 'person'}, {type: 'post'}]}],
        },
        {type: 'reference', title: 'Reference', to: [{type: 'person'}, {type: 'post'}]},
        {type: 'image'},
        {type: 'file'},
      ],
      title: 'biography',
      hidden: false,
    }),
    defineField({
      name: 'friend',
      type: 'reference',
      title: 'friend',
      hidden: false,
      to: [{type: 'person'}, {type: 'post'}],
    }),
    defineField({
      name: 'friends',
      type: 'array',
      of: [{type: 'reference', to: [{type: 'person'}, {type: 'post'}]}],
      title: 'friends',
      hidden: false,
    }),
    defineField({
      type: 'boolean',
      description:
        'If this document was archived on Contentful at the time of export, the document will be in a read-only state.',
      name: 'contentfulArchived',
      readOnly: true,
    }),
  ],
  preview: {select: {title: 'name'}},
  readOnly: ({document}) => document?.contentfulArchived === true,
})

export const postType = defineType({
  type: 'document',
  name: 'post',
  title: 'Post',
  description: '',
  fields: [
    defineField({name: 'title', type: 'string', title: 'title', hidden: false}),
    defineField({
      name: 'author',
      type: 'reference',
      title: 'author',
      hidden: false,
      to: [{type: 'person'}, {type: 'post'}],
    }),
    defineField({
      name: 'body',
      type: 'array',
      of: [
        {
          type: 'block',
          styles: [
            {title: 'Heading 1', value: 'h1'},
            {title: 'Heading 2', value: 'h2'},
            {title: 'Heading 3', value: 'h3'},
            {title: 'Heading 4', value: 'h4'},
            {title: 'Heading 5', value: 'h5'},
            {title: 'Heading 6', value: 'h6'},
            {title: 'Quote', value: 'blockquote'},
          ],
          lists: [
            {title: 'Bullet', value: 'bullet'},
            {title: 'Numbered', value: 'number'},
          ],
          marks: {
            decorators: [
              {title: 'Strong', value: 'strong'},
              {title: 'Emphasis', value: 'em'},
              {title: 'Underline', value: 'underline'},
              {title: 'Code', value: 'pre'},
            ],
            annotations: [
              {
                type: 'object',
                name: 'link',
                title: 'url',
                fields: [
                  defineField({
                    type: 'string',
                    name: 'href',
                    title: 'URL',
                    validation: (Rule) => Rule.required(),
                  }),
                  defineField({
                    type: 'string',
                    name: 'target',
                    title: 'Target',
                    options: {
                      list: [
                        {value: '_blank', title: 'Blank'},
                        {value: '_parent', title: 'Parent'},
                      ],
                    },
                  }),
                ],
              },
              {
                type: 'object',
                name: 'internalLink',
                title: 'Internal link',
                fields: [
                  defineField({
                    type: 'reference',
                    name: 'reference',
                    to: [{type: 'person'}, {type: 'post'}],
                  }),
                ],
              },
              {type: 'image'},
              {type: 'file'},
            ],
          },
          of: [{type: 'reference', title: 'Reference', to: [{type: 'person'}, {type: 'post'}]}],
        },
        {type: 'reference', title: 'Reference', to: [{type: 'person'}, {type: 'post'}]},
        {type: 'image'},
        {type: 'file'},
        {type: 'break'},
      ],
      title: 'body',
      hidden: false,
    }),
    defineField({
      type: 'boolean',
      description:
        'If this document was archived on Contentful at the time of export, the document will be in a read-only state.',
      name: 'contentfulArchived',
      readOnly: true,
    }),
  ],
  preview: {select: {title: 'title'}},
  readOnly: ({document}) => document?.contentfulArchived === true,
})

export const breakType = defineType({
  name: 'break',
  title: 'Break',
  type: 'object',
  fields: [
    defineField({
      name: 'style',
      type: 'string',
      options: {
        list: [
          {title: 'Line break', value: 'lineBreak'},
          {title: 'Read more', value: 'readMore'},
        ],
      },
    }),
  ],
})

export const types = [personType, postType, breakType] satisfies SchemaTypeDefinition[]
"
`;
